### PR TITLE
ADEN-7078 SPA support for AE3

### PR DESCRIPTION
**Pull request by leshniak**
## Links

* http://wikia-inc.atlassian.net/browse/ADEN-7078

## Description

SPA support for AdEngine3

### CODE REVIEW

File: app/routes/application.js [hunk 1]
```diff
@@ -93,14 +93,14 @@ export default Route.extend(
 			) {
 
 				window.waitForAds((adEngine3Loaded) => {
+					const adsModule = this.get('ads.module');
+
+					adsModule.init(model.adsContext);
```
Comment (fraszczakszymon) [inline app/routes/application.js]: There is no adsContext at this point. AE3 has to be initialised in wiki-page.js

https://github.com/Wikia/mobile-wiki/pull/621/files#diff-d77b3e7a178bee9c8affd145b2261a54R131

Comment (fraszczakszymon) [inline app/routes/application.js]: It should't happen here for adEngine3 here (move it below if statement please)

File: app/routes/wiki-page.js [hunk 2]
```diff
@@ -126,10 +126,8 @@ export default Route.extend(
 						!fastboot.get('isFastBoot') &&
 						!transition.queryParams.noexternals
 					) {
-						window.waitForAds((adEngine3Loaded) => {
-							if (adEngine3Loaded) {
-								this.get('ads.module').setupAdEngine3_ThisMethodShouldBeDefinedInOtherModule(model.adsContext);
-							}
+						window.waitForAds(() => {
+							this.get('ads.module').init(model.adsContext);
```
Comment (fraszczakszymon) [inline app/routes/wiki-page.js]: ads.module.init should be executed for ae3 only (if statement was fine)

File: vendor/inline-scripts/load-ads.js [hunk 3]
```diff
@@ -27,9 +23,13 @@
 	}
 
 	window.getInstantGlobals(function (instantGlobals) {
-		var adsDisabled = instantGlobals.wgSitewideDisableAdsOnMercury
+		var noExternalsSearchParam = (window.location.search.match(/noexternals=([a-z0-9]+)/i) || [])[1];
```
Comment (nandy-andy) [inline vendor/inline-scripts/load-ads.js:26]: What is the reason for using regex instead previously used `URL(document.location).searchParams.get('noexternals')`? :-)

File: vendor/inline-scripts/load-ads.js [hunk 4]
```diff
@@ -40,27 +40,23 @@
...
-				.then(function () {
-					return loadScript(assetUrls.adProductsScript)
-				})
-				.then(function () {
-					return loadScript(assetUrls.geoScript)
-				})
-				.then(function () {
-					adEngine3Loaded = true;
-					onAdsLoaded();
+			loadScript(assetUrls.adEngineScript, function onEngineLoaded() {
```
Comment (nandy-andy) [inline vendor/inline-scripts/load-ads.js:43]: What was wrong with promises? ;-)

File: app/modules/ads/module.js [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+	}
+
+	/**
+	 * Turns off all ads for logged in user
+	 *
+	 * @param {*} adsContext
+	 * @returns {void}
+	 */
+	turnOffAdsForLoggedInUsers(adsContext) {
+		// TODO: Refactor/remove while working on ADEN-2189
```
Comment (nandy-andy) [inline app/modules/ads/module.js:285]: ADEN-2189 is a won't fix. let's just face it - we won't do it until it's needed. let's remove the TODO.

File: app/modules/ads/setup.js [hunk 6]
```diff
@@ -121,11 +118,15 @@ function init() {
...
+	engine.init();
+
+	return engine;
 }
 
 export default {
 	configure,
 	init,
-	setupAdContext
+	setupAdContext // TODO: re-setupAdContext on page transition (with new mediaWikiContext)
```
Comment (nandy-andy) [inline app/modules/ads/setup.js:134]: When are we going to implement re-setupAdContext?

File: vendor/inline-scripts/load-ads.js [hunk 7]
```diff
@@ -40,27 +40,23 @@
 		// Note that at this point we don't have geo module
 		// TODO#2: load adProducts* script in parallel
```
Comment (nandy-andy) [inline vendor/inline-scripts/load-ads.js:41]: When are we going to address TODOs?

File: vendor/inline-scripts/load-ads.js (hunk 3 above)
Comment (dmnsobczak) [reply vendor/inline-scripts/load-ads.js:26]: It's related to recent iOS9 fix: https://github.com/Wikia/mobile-wiki/pull/658/files

File: app/modules/ads/module.js (hunk 5 above)
Comment (leshniak) [reply app/modules/ads/module.js:285]: ok, chef.

File: vendor/inline-scripts/load-ads.js (hunk 3 above)
Comment (leshniak) [reply vendor/inline-scripts/load-ads.js:26]: yup, exactly

File: vendor/inline-scripts/load-ads.js (hunk 4 above)
Comment (leshniak) [reply vendor/inline-scripts/load-ads.js:43]: This file is not transpiled. It's for backward compatibility.

**Pull request merged**